unexpected error when unsubscribe service merge2,merge3/org.apache.dubbo.samples.merge.api.MergeService from registry: nacos, dubbo version 3.1.0

```java
[03/09/22 06:27:41:041 CST] DubboShutdownHook  WARN integration.DynamicDirectory:  [DUBBO] unexpected error when unsubscribe service merge2,merge3/org.apache.dubbo.samples.merge.api.MergeService from registry: nacos://127.0.0.1:8848/org.apache.dubbo.registry.RegistryService?application=merge-consumer&dubbo=2.0.2&interface=org.apache.dubbo.registry.RegistryService&pid=26840&release=3.1.0, dubbo version: 3.1.0, current host: 30.221.208.119, error code: 1-8. This may be caused by , go to https://dubbo.apache.org/faq/1/8 to find instructions. 
java.lang.IllegalStateException: Failed to unsubscribe consumer://30.221.208.119/org.apache.dubbo.samples.merge.api.MergeService?application=merge-consumer&background=false&category=providers,configurators,routers&dubbo=2.0.2&group=merge2,merge3&interface=org.apache.dubbo.samples.merge.api.MergeService&methods=mergeResult&pid=26840&release=3.1.0&side=consumer&sticky=false&timestamp=1662200561561 to registry 127.0.0.1:8848, cause: null
	at org.apache.dubbo.registry.support.FailbackRegistry.unsubscribe(FailbackRegistry.java:344)
	at org.apache.dubbo.registry.ListenerRegistryWrapper.unsubscribe(ListenerRegistryWrapper.java:136)
	at org.apache.dubbo.registry.integration.DynamicDirectory.destroy(DynamicDirectory.java:309)
	at org.apache.dubbo.registry.integration.RegistryDirectory.destroy(RegistryDirectory.java:135)
	at org.apache.dubbo.rpc.cluster.support.AbstractClusterInvoker.destroy(AbstractClusterInvoker.java:128)
	at org.apache.dubbo.rpc.cluster.support.wrapper.MockClusterInvoker.destroy(MockClusterInvoker.java:87)
	at org.apache.dubbo.registry.client.migration.MigrationInvoker.destroy(MigrationInvoker.java:325)
	at org.apache.dubbo.rpc.proxy.InvokerInvocationHandler.invoke(InvokerInvocationHandler.java:61)
	at org.apache.dubbo.samples.merge.api.MergeServiceDubboProxy1.$destroy(MergeServiceDubboProxy1.java)
	at org.apache.dubbo.config.utils.SimpleReferenceCache.destroyReference(SimpleReferenceCache.java:264)
	at org.apache.dubbo.config.utils.SimpleReferenceCache.lambda$destroyAll$4(SimpleReferenceCache.java:253)
	at java.util.concurrent.ConcurrentHashMap.forEach(ConcurrentHashMap.java:1597)
	at org.apache.dubbo.config.utils.SimpleReferenceCache.destroyAll(SimpleReferenceCache.java:251)
	at org.apache.dubbo.config.deploy.DefaultModuleDeployer.unreferServices(DefaultModuleDeployer.java:402)
	at org.apache.dubbo.config.deploy.DefaultModuleDeployer.postDestroy(DefaultModuleDeployer.java:208)
	at org.apache.dubbo.rpc.model.ModuleModel.onDestroy(ModuleModel.java:108)
	at org.apache.dubbo.rpc.model.ScopeModel.destroy(ScopeModel.java:113)
	at org.apache.dubbo.rpc.model.ApplicationModel.onDestroy(ApplicationModel.java:256)
	at org.apache.dubbo.rpc.model.ScopeModel.destroy(ScopeModel.java:113)
	at org.apache.dubbo.config.DubboShutdownHook.doDestroy(DubboShutdownHook.java:78)
	at org.apache.dubbo.config.DubboShutdownHook.run(DubboShutdownHook.java:73)
Caused by: java.lang.NullPointerException
	at org.apache.dubbo.registry.nacos.NacosRegistry.doUnsubscribe(NacosRegistry.java:280)
	at org.apache.dubbo.registry.support.FailbackRegistry.unsubscribe(FailbackRegistry.java:332)
	... 20 more
```

--- a/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java
+++ b/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java
@@ -287,7 +287,7 @@ public class NacosRegistry extends FailbackRegistry {
                     doUnsubscribe(url, nacosAggregateListener, serviceNames);
                     NacosInstanceManageUtil.removeSubscribedServiceNames(url.getServiceKey());
                     for (String serviceName : serviceNames) {
-                        NacosInstanceManageUtil.removeSubscribedServiceNames(serviceName);
+                        NacosInstanceManageUtil.getCorrespondingServiceNames(serviceName);
                     }
                 } catch (NacosException e) {
                     logger.error("Failed to unsubscribe " + url + " to nacos " + getUrl() + ", cause: " + e.getMessage(), e);

--- a/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java
+++ b/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java
@@ -214,8 +214,6 @@ public class NacosRegistry extends FailbackRegistry {
                 NacosInstanceManageUtil.setCorrespondingServiceNames(serviceName, serviceNames);
             }
         }
-        // save subscribed dataIds for unsubscription usage later
-        NacosInstanceManageUtil.setSubscribedServiceNames(url.getServiceKey(), serviceNames);
 
         doSubscribe(url, nacosAggregateListener, serviceNames);
     }
@@ -280,12 +278,15 @@ public class NacosRegistry extends FailbackRegistry {
             shutdownServiceNamesLookup();
         } else {
             Map<NotifyListener, NacosAggregateListener> listenerMap = originToAggregateListener.get(url);
+            if (listenerMap == null) {
+                logger.warn(String.format("No aggregate listener found for url %s, this service might have already been unsubscribed.", url));
+                return;
+            }
             NacosAggregateListener nacosAggregateListener = listenerMap.remove(listener);
             if (nacosAggregateListener != null) {
-                Set<String> serviceNames = NacosInstanceManageUtil.getSubscribedServiceNames(url.getServiceKey());
+                Set<String> serviceNames = nacosAggregateListener.getServiceNames();
                 try {
                     doUnsubscribe(url, nacosAggregateListener, serviceNames);
-                    NacosInstanceManageUtil.removeSubscribedServiceNames(url.getServiceKey());
                     for (String serviceName : serviceNames) {
                         NacosInstanceManageUtil.getCorrespondingServiceNames(serviceName);
                     }

--- a/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java
+++ b/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java
@@ -214,6 +214,9 @@ public class NacosRegistry extends FailbackRegistry {
                 NacosInstanceManageUtil.setCorrespondingServiceNames(serviceName, serviceNames);
             }
         }
+        // save subscribed dataIds for unsubscription usage later
+        NacosInstanceManageUtil.setSubscribedServiceNames(url.getServiceKey(), serviceNames);
+
         doSubscribe(url, nacosAggregateListener, serviceNames);
     }
 
@@ -279,9 +282,13 @@ public class NacosRegistry extends FailbackRegistry {
             Map<NotifyListener, NacosAggregateListener> listenerMap = originToAggregateListener.get(url);
             NacosAggregateListener nacosAggregateListener = listenerMap.remove(listener);
             if (nacosAggregateListener != null) {
-                Set<String> serviceNames = getServiceNames(url, nacosAggregateListener);
+                Set<String> serviceNames = NacosInstanceManageUtil.getSubscribedServiceNames(url.getServiceKey());
                 try {
                     doUnsubscribe(url, nacosAggregateListener, serviceNames);
+                    NacosInstanceManageUtil.removeSubscribedServiceNames(url.getServiceKey());
+                    for (String serviceName : serviceNames) {
+                        NacosInstanceManageUtil.removeSubscribedServiceNames(serviceName);
+                    }
                 } catch (NacosException e) {
                     logger.error("Failed to unsubscribe " + url + " to nacos " + getUrl() + ", cause: " + e.getMessage(), e);
                 }

--- a/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java
+++ b/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java
@@ -277,7 +277,7 @@ public class NacosRegistry extends FailbackRegistry {
         if (isAdminProtocol(url)) {
             shutdownServiceNamesLookup();
         } else {
-                Map<NotifyListener, NacosAggregateListener> listenerMap = originToAggregateListener.get(url);
+            Map<NotifyListener, NacosAggregateListener> listenerMap = originToAggregateListener.get(url);
             if (listenerMap == null) {
                 logger.warn(String.format("No aggregate listener found for url %s, this service might have already been unsubscribed.", url));
                 return;

--- a/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java
+++ b/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java
@@ -277,7 +277,7 @@ public class NacosRegistry extends FailbackRegistry {
         if (isAdminProtocol(url)) {
             shutdownServiceNamesLookup();
         } else {
-            Map<NotifyListener, NacosAggregateListener> listenerMap = originToAggregateListener.get(url);
+                Map<NotifyListener, NacosAggregateListener> listenerMap = originToAggregateListener.get(url);
             if (listenerMap == null) {
                 logger.warn(String.format("No aggregate listener found for url %s, this service might have already been unsubscribed.", url));
                 return;
@@ -288,7 +288,7 @@ public class NacosRegistry extends FailbackRegistry {
                 try {
                     doUnsubscribe(url, nacosAggregateListener, serviceNames);
                     for (String serviceName : serviceNames) {
-                        NacosInstanceManageUtil.getCorrespondingServiceNames(serviceName);
+                        NacosInstanceManageUtil.removeCorrespondingServiceNames(serviceName);
                     }
                 } catch (NacosException e) {
                     logger.error("Failed to unsubscribe " + url + " to nacos " + getUrl() + ", cause: " + e.getMessage(), e);

--- a/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/util/NacosInstanceManageUtil.java
+++ b/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/util/NacosInstanceManageUtil.java
@@ -44,23 +44,6 @@ public class NacosInstanceManageUtil {
      */
     private static final Map<String, Set<String>> CORRESPONDING_SERVICE_NAMES_MAP = Maps.newConcurrentMap();
 
-    /**
-     * service key -> subscribed serviceName list
-     */
-    private static final Map<String, Set<String>> SUBSCRIBED_SERVICE_NAMES_MAP = Maps.newConcurrentMap();
-
-    public static void setSubscribedServiceNames(String serviceKey, Set<String> serviceNames) {
-        SUBSCRIBED_SERVICE_NAMES_MAP.put(serviceKey, serviceNames);
-    }
-
-    public static Set<String> getSubscribedServiceNames(String removeSubscribedServiceNames) {
-        return SUBSCRIBED_SERVICE_NAMES_MAP.get(removeSubscribedServiceNames);
-    }
-
-    public static Set<String> removeSubscribedServiceNames(String serviceKey) {
-        return SUBSCRIBED_SERVICE_NAMES_MAP.remove(serviceKey);
-    }
-
     public static void setCorrespondingServiceNames(String serviceName, Set<String> serviceNames) {
         CORRESPONDING_SERVICE_NAMES_MAP.put(serviceName, serviceNames);
     }
@@ -74,7 +57,7 @@ public class NacosInstanceManageUtil {
     }
 
     public static Set<String> removeCorrespondingServiceNames(String serviceName) {
-        return SUBSCRIBED_SERVICE_NAMES_MAP.remove(serviceName);
+        return CORRESPONDING_SERVICE_NAMES_MAP.remove(serviceName);
     }
 
     public static List<Instance> getAllCorrespondingServiceInstanceList(String serviceName) {

--- a/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/util/NacosInstanceManageUtil.java
+++ b/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/util/NacosInstanceManageUtil.java
@@ -44,12 +44,37 @@ public class NacosInstanceManageUtil {
      */
     private static final Map<String, Set<String>> CORRESPONDING_SERVICE_NAMES_MAP = Maps.newConcurrentMap();
 
+    /**
+     * service key -> subscribed serviceName list
+     */
+    private static final Map<String, Set<String>> SUBSCRIBED_SERVICE_NAMES_MAP = Maps.newConcurrentMap();
+
+    public static void setSubscribedServiceNames(String serviceKey, Set<String> serviceNames) {
+        SUBSCRIBED_SERVICE_NAMES_MAP.put(serviceKey, serviceNames);
+    }
+
+    public static Set<String> getSubscribedServiceNames(String removeSubscribedServiceNames) {
+        return SUBSCRIBED_SERVICE_NAMES_MAP.get(removeSubscribedServiceNames);
+    }
+
+    public static Set<String> removeSubscribedServiceNames(String serviceKey) {
+        return SUBSCRIBED_SERVICE_NAMES_MAP.remove(serviceKey);
+    }
+
     public static void setCorrespondingServiceNames(String serviceName, Set<String> serviceNames) {
         CORRESPONDING_SERVICE_NAMES_MAP.put(serviceName, serviceNames);
     }
 
+    public static Set<String> getCorrespondingServiceNames(String serviceName) {
+        return CORRESPONDING_SERVICE_NAMES_MAP.get(serviceName);
+    }
+
     public static void initOrRefreshServiceInstanceList(String serviceName, List<Instance> instanceList) {
         SERVICE_INSTANCE_LIST_MAP.put(serviceName, instanceList);
+    }
+
+    public static Set<String> removeCorrespondingServiceNames(String serviceName) {
+        return SUBSCRIBED_SERVICE_NAMES_MAP.remove(serviceName);
     }
 
     public static List<Instance> getAllCorrespondingServiceInstanceList(String serviceName) {

--- a/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/util/NacosInstanceManageUtil.java
+++ b/dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/util/NacosInstanceManageUtil.java
@@ -48,10 +48,6 @@ public class NacosInstanceManageUtil {
         CORRESPONDING_SERVICE_NAMES_MAP.put(serviceName, serviceNames);
     }
 
-    public static Set<String> getCorrespondingServiceNames(String serviceName) {
-        return CORRESPONDING_SERVICE_NAMES_MAP.get(serviceName);
-    }
-
     public static void initOrRefreshServiceInstanceList(String serviceName, List<Instance> instanceList) {
         SERVICE_INSTANCE_LIST_MAP.put(serviceName, instanceList);
     }

